MvvmCross/Core/Binding/BindingContext/MvxTaskBasedBindingContext.cs

```diff
@@ -0,0 +1,209 @@
+using MvvmCross.Platform;
+
+namespace MvvmCross.Binding.BindingContext
+{
+    /// <summary>
+    /// OnDataContextChange executes asynchronously on a worker thread
+    /// </summary>
+    public class MvxTaskBasedBindingContext : IMvxBindingContext
+    {
+        private readonly List<Action> _delayedActions = new List<Action>();
+        private readonly List<MvxBindingContext.TargetAndBinding> _directBindings = new List<MvxBindingContext.TargetAndBinding>();
+        private readonly List<KeyValuePair<object, IList<MvxBindingContext.TargetAndBinding>>> _viewBindings = new List<KeyValuePair<object, IList<MvxBindingContext.TargetAndBinding>>>();
+        private object _dataContext;
+        private IMvxBinder _binder;
+
+        public event EventHandler DataContextChanged;
+
+        public IMvxBindingContext Init(object dataContext, object firstBindingKey, IEnumerable<MvxBindingDescription> firstBindingValue)
+        {
+            this.AddDelayedAction(firstBindingKey, firstBindingValue);
+            if (dataContext != null)
+                this.DataContext = dataContext;
+
+            return this;
+        }
+
+        public IMvxBindingContext Init(object dataContext, object firstBindingKey, string firstBindingValue)
+        {
+            this.AddDelayedAction(firstBindingKey, firstBindingValue);
+            if (dataContext != null)
+                this.DataContext = dataContext;
+
+            return this;
+        }
+
+        private void AddDelayedAction(object key, string value)
+        {
+            this._delayedActions.Add(() =>
+            {
+                var bindings = this.Binder.Bind(this.DataContext, key, value);
+                foreach (var b in bindings)
+                    this.RegisterBinding(key, b);
+            });
+        }
+
+        private void AddDelayedAction(object key, IEnumerable<MvxBindingDescription> value)
+        {
+            this._delayedActions.Add(() =>
+            {
+                var bindings = this.Binder.Bind(this.DataContext, key, value);
+                foreach (var b in bindings)
+                    this.RegisterBinding(key, b);
+            });
+        }
+
+        ~MvxTaskBasedBindingContext()
+        {
+            this.Dispose(false);
+        }
+
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                this.ClearAllBindings();
+            }
+        }
+
+        protected IMvxBinder Binder
+        {
+            get
+            {
+                this._binder = this._binder ?? Mvx.Resolve<IMvxBinder>();
+                return this._binder;
+            }
+        }
+
+        public object DataContext
+        {
+            get { return this._dataContext; }
+            set
+            {
+                this._dataContext = value;
```

Comment:
As in `if (this._dataContext == value) return;`.  Not a big deal if that's not the pattern.